File: ecommerce-web-admin/src/services/transactionService.ts
import { 
  collection, 
  query, 
  where, 
  orderBy, 
  getDocs, 
  doc, 
  updateDoc,
  getCountFromServer
} from 'firebase/firestore';
import { db } from './firebase';
import { Transaction } from '../types';

export class TransactionService {
  
  // Get all transactions
  async getAllTransactions(): Promise<Transaction[]> {
    try {
      const transactionsSnapshot = await getDocs(
        query(collection(db, 'transactions'), orderBy('timestamp', 'desc'))
      );
      
      return transactionsSnapshot.docs.map(doc => ({
        id: doc.id,
        ...doc.data()
      } as Transaction));
    } catch (error) {
      console.error('Error getting all transactions:', error);
      return [];
    }
  }

  // Get transactions by status
  async getTransactionsByStatus(status: string): Promise<Transaction[]> {
    try {
      const transactionsQuery = query(
        collection(db, 'transactions'),
        where('status', '==', status),
        orderBy('timestamp', 'desc')
      );
      const transactionsSnapshot = await getDocs(transactionsQuery);
      
      return transactionsSnapshot.docs.map(doc => ({
        id: doc.id,
        ...doc.data()
      } as Transaction));
    } catch (error) {
      console.error('Error getting transactions by status:', error);
      return [];
    }
  }

  // Get transactions by date range
  async getTransactionsByDateRange(startDate: Date, endDate: Date): Promise<Transaction[]> {
    try {
      const transactionsQuery = query(
        collection(db, 'transactions'),
        where('timestamp', '>=', startDate),
        where('timestamp', '<=', endDate),
        orderBy('timestamp', 'desc')
      );
      const transactionsSnapshot = await getDocs(transactionsQuery);
      
      return transactionsSnapshot.docs.map(doc => ({
        id: doc.id,
        ...doc.data()
      } as Transaction));
    } catch (error) {
      console.error('Error getting transactions by date range:', error);
      return [];
    }
  }

  // Update transaction status
  async updateTransactionStatus(transactionId: string, status: string): Promise<boolean> {
    try {
      await updateDoc(doc(db, 'transactions', transactionId), {
        status
      });
      return true;
    } catch (error) {
      console.error('Error updating transaction status:', error);
      return false;
    }
  }

  // Get transaction stats
  async getTransactionStats(): Promise<{ 
    totalTransactions: number; 
    completedTransactions: number; 
    pendingTransactions: number;
    totalRevenue: number;
  }> {
    try {
      // Count total transactions
      const totalSnapshot = await getCountFromServer(collection(db, 'transactions'));
      const totalTransactions = totalSnapshot.data().count;

      // Count completed transactions
      const completedTransactionsQuery = query(
        collection(db, 'transactions'),
        where('status', '==', 'completed')
      );
      const completedTransactionsSnapshot = await getCountFromServer(completedTransactionsQuery);
      const completedTransactions = completedTransactionsSnapshot.data().count;

      // Count pending transactions
      const pendingTransactionsQuery = query(
        collection(db, 'transactions'),
        where('status', '==', 'pending')
      );
      const pendingTransactionsSnapshot = await getCountFromServer(pendingTransactionsQuery);
      const pendingTransactions = pendingTransactionsSnapshot.data().count;

      // Calculate total revenue (from completed transactions)
      const completedTransactionsDocs = await getDocs(completedTransactionsQuery);
      const totalRevenue = completedTransactionsDocs.docs.reduce((sum, doc) => {
        const data = doc.data();
        return sum + (data.amount || 0);
      }, 0);

      return {
        totalTransactions,
        completedTransactions,
        pendingTransactions,
        totalRevenue
      };
    } catch (error) {
      console.error('Error getting transaction stats:', error);
      return {
        totalTransactions: 0,
        completedTransactions: 0,
        pendingTransactions: 0,
        totalRevenue: 0
      };
    }
  }
}

File: e-commerce-app/lib/services/transaction_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';

class TransactionService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // Get all transactions
  Future<List<Map<String, dynamic>>> getAllTransactions() async {
    try {
      QuerySnapshot querySnapshot = await _firestore.collection('transactions').get();
      return querySnapshot.docs.map((doc) {
        Map<String, dynamic> data = doc.data() as Map<String, dynamic>;
        data['id'] = doc.id;
        return data;
      }).toList();
    } catch (e) {
      print('Error getting all transactions: $e');
      return [];
    }
  }

  // Get transactions by status (pending, completed, canceled)
  Future<List<Map<String, dynamic>>> getTransactionsByStatus(String status) async {
    try {
      QuerySnapshot querySnapshot = await _firestore
          .collection('transactions')
          .where('status', isEqualTo: status)
          .get();
      
      return querySnapshot.docs.map((doc) {
        Map<String, dynamic> data = doc.data() as Map<String, dynamic>;
        data['id'] = doc.id;
        return data;
      }).toList();
    } catch (e) {
      print('Error getting transactions by status: $e');
      return [];
    }
  }

  // Get transactions by date range
  Future<List<Map<String, dynamic>>> getTransactionsByDateRange(DateTime startDate, DateTime endDate) async {
    try {
      // Add one day to end date to include the whole end date
      DateTime endDatePlusOne = endDate.add(const Duration(days: 1));
      
      // Convert DateTime to ISO string for Firestore
      String startDateStr = startDate.toIso8601String();
      String endDatePlusOneStr = endDatePlusOne.toIso8601String();
      
      QuerySnapshot querySnapshot = await _firestore
          .collection('transactions')
          .where('createdAt', isGreaterThanOrEqualTo: startDateStr)
          .where('createdAt', isLessThan: endDatePlusOneStr)
          .get();
      
      return querySnapshot.docs.map((doc) {
        Map<String, dynamic> data = doc.data() as Map<String, dynamic>;
        data['id'] = doc.id;
        return data;
      }).toList();
    } catch (e) {
      print('Error getting transactions by date range: $e');
      return [];
    }
  }

  // Get transactions by user
  Future<List<Map<String, dynamic>>> getTransactionsByUser(String userId) async {
    try {
      QuerySnapshot querySnapshot = await _firestore
          .collection('transactions')
          .where('userId', isEqualTo: userId)
          .get();
      
      return querySnapshot.docs.map((doc) {
        Map<String, dynamic> data = doc.data() as Map<String, dynamic>;
        data['id'] = doc.id;
        return data;
      }).toList();
    } catch (e) {
      print('Error getting transactions by user: $e');
      return [];
    }
  }
  
  // Get transaction stats
  Future<Map<String, int>> getTransactionStats() async {
    try {
      AggregateQuerySnapshot totalSnapshot = await _firestore.collection('transactions').count().get();
      AggregateQuerySnapshot completedSnapshot = await _firestore
          .collection('transactions')
          .where('status', isEqualTo: 'completed')
          .count()
          .get();
      
      // Handle potentially nullable counts with null-aware operators
      int totalTransactions = totalSnapshot.count ?? 0;
      int completedTransactions = completedSnapshot.count ?? 0;
      
      return {
        'totalTransactions': totalTransactions,
        'completedTransactions': completedTransactions,
      };
    } catch (e) {
      print('Error getting transaction stats: $e');
      return {
        'totalTransactions': 0,
        'completedTransactions': 0,
      };
    }
  }
  
  // Get weekly transaction activity (for graph)
  Future<Map<String, double>> getWeeklyTransactionActivity() async {
    try {
      // Get current date
      DateTime now = DateTime.now();
      
      // Create a map to store data for the last 7 days
      Map<String, double> weeklyActivity = {};
      
      // Populate map with last 7 days (including today)
      for (int i = 6; i >= 0; i--) {
        DateTime date = now.subtract(Duration(days: i));
        String dateString = '${date.year}-${date.month.toString().padLeft(2, '0')}-${date.day.toString().padLeft(2, '0')}';
        weeklyActivity[dateString] = 0.0;
      }
      
      // Query transactions created in the last 7 days
      DateTime sevenDaysAgo = now.subtract(const Duration(days: 7));
      String sevenDaysAgoStr = sevenDaysAgo.toIso8601String();
      
      QuerySnapshot querySnapshot = await _firestore
          .collection('transactions')
          .where('createdAt', isGreaterThanOrEqualTo: sevenDaysAgoStr)
          .get();
          
      // Sum transaction amounts by day
      for (var doc in querySnapshot.docs) {
        final data = doc.data() as Map<String, dynamic>;
        if (data['createdAt'] != null) {
          DateTime? createdAt;
          
          // Handle different timestamp formats
          if (data['createdAt'] is String) {
            try {
              createdAt = DateTime.parse(data['createdAt']);
            } catch (e) {
              print('Error parsing transaction date string: ${data['createdAt']}');
              continue;
            }
          } else if (data['createdAt'].runtimeType.toString().contains('Timestamp')) {
            // Firestore Timestamp
            createdAt = data['createdAt'].toDate();
          }
          
          if (createdAt != null) {
            String dateString = '${createdAt.year}-${createdAt.month.toString().padLeft(2, '0')}-${createdAt.day.toString().padLeft(2, '0')}';
            
            // Safely convert to double, handling various types
            double amount = 0.0;
            var rawAmount = data['amount'];
            if (rawAmount is int) {
              amount = rawAmount.toDouble();
            } else if (rawAmount is double) {
              amount = rawAmount;
            } else if (rawAmount != null) {
              amount = double.tryParse(rawAmount.toString()) ?? 0.0;
            }
            
            if (weeklyActivity.containsKey(dateString)) {
              weeklyActivity[dateString] = weeklyActivity[dateString]! + amount;
            }
          }
        }
      }
      
      return weeklyActivity;
    } catch (e) {
      print('Error getting weekly transaction activity: $e');
      return {};
    }
  }
}
